### PR TITLE
Did you mean to block nightlies on clippy?

### PR DESCRIPTION
**Pull request by oli-obk**
Discussion: https://gitter.im/rust-lang/WG-clippy?at=5b073b6597a0361fb760cdc2

r? @alexcrichton 

did I forget anything?

cc @nrc @Manishearth

### CODE REVIEW

**Comment by kennytm** [comment]
Did you *really* mean to block nightlies on clippy? 😏 

* If you want clippy to block nightlies, you need to submit a PR to rust-central-station later (like rust-lang/rust-central-station@913a3b3650231884a6bc8ad3cd98a44e8552b45a). 
* If you want clippy to block betas and stables, you need to update `src/ci/docker/x86_64-gnu-tools/checktools.sh`.

**Comment by oli-obk** [comment]
> If you want clippy to block betas and stables

We'll first see how this is going

File: src/bootstrap/dist.rs [hunk 1]
```diff
@@ -40,6 +40,8 @@ pub fn pkgname(builder: &Builder, component: &str) -> String {
     if component == "cargo" {
         format!("{}-{}", component, builder.cargo_package_vers())
     } else if component == "rls" {
+        format!("{}-{}", component, builder.clippy_package_vers())
```
Comment (Keruspe) [inline src/bootstrap/dist.rs]: You swapped rls and clippy here

Comment (oli-obk) [reply src/bootstrap/dist.rs]: thanks! Copy paste at its worst

**Comment by nrc** [comment]
To clarify, I meant that I think we should do this in a few weeks once we have collected some data about Clippy tests failing and had a discussion with the dev-tools and core teams.

**Comment by oli-obk** [comment]
Oh... I understood that there had been this discussion and we should do this change in order to collect said data.

Apropos, this PR does not actually block nightlies on clippy, it just ships the clippy-driver binary. With a tiny change I can make this PR not produce a component. If we then merge this PR we'll have much easier time in adding clippy later.

So... how do we collect that data? Do I just give you a histogram of number of hours from clippy breaking to clippy releasing a new version? Or do you want the opposite? A histogram for how long clippy actually works? Because we have this data in some form iin the clippy repo.

**Comment by oli-obk** [comment]
I just talked with @Manishearth about this. Here's what I got out of that discussion:

1. The data we have is that usually clippy breaks around once every two weeks. Sometimes though we get 3-4 days where it breaks every day, and then there's a period where it works for a while.

2. We don't often fix clippy ahead of time (before a nightly hits that actually breaks clippy), because there's little use in doing so. If nightlies were blocked on clippy, we'd fix clippy and update the submodule immediately when it's broken (the fixes in the last few months were trivial renamings and some restructurings, nothing where clippy has to figure out a new way to do something)

3. cc @rust-lang/infra would it make sense to speed up merging submodule-only changes by taking the binaries of the previously merged PR and just testing the changed submodules?

**Comment by Manishearth** [comment]
We probably should get accurate data on this, fwiw, this is what I've seen
from rough measurements.

**Comment by oli-obk** [comment]
I generated a diagram from our CHANGELOG.md:

grafik

X is the number of days between clippy versions. The same diagram for just 2018 is

grafik

So it seems that the majority of no breakage happening is in the 3-5 day range

Note that this data is obviously wrong, because if we don't release a new version quickly, the number of days between releases goes up, even if the breakage happens after fewer days.
Another reason this data is wrong is due to the fact that as long as rls is broken, we're not getting new nightlies, and thus clippy isn't breaking.

**Comment by kennytm** [comment]
@oli-obk 

I'm not sure if it's worth it.  The main difficulty is determine what is meant by "submodule-only changes", since often you also need to update `Cargo.lock`, and *that* may change the behavior of the rest of `rustc`.

**Comment by shepmaster** [comment]
Ping from triage, @nrc !

**Comment by nrc** [comment]
Nominating for discussion in this weeks meeting

**Comment by Manishearth** [comment]
We discussed this today's meeting, we'd like to merge this, _without_ blocking nightlies (adding a component). Hopefully we can move to a component based model for clippy from here, and eventually block nightlies. The Rustup support for this exists but may not be perfectly tested, so we may have some stuff to fix here.

One thing we also discussed was the data. It seems like one of the problems is that if clippy breaks nightly, it takes long enough to fix it that other breakages end up entering the queue. For this, we should:

 - p=100 component update PRs
 - Use https://github.com/kennytm/rustup-toolchain-install-master so we don't have to build rustc
 - Probably maintain a temporary branch on clippy and avoid doing releases.

r? @kennytm on the toolchain code?

File: src/bootstrap/dist.rs [hunk 2]
```diff
@@ -1178,6 +1180,84 @@ impl Step for Rls {
...
+        let clippy = builder.ensure(tool::Clippy {
+            compiler: builder.compiler(stage, builder.config.build),
+            target, extra_features: Vec::new()
+        }).or_else(|| { println!("Unable to build clippy, skipping dist"); None })?;
+
+        builder.install(&clippy, &image.join("bin"), 0o755);
+        let doc = image.join("share/doc/clippy");
+        builder.install(&src.join("README.md"), &doc, 0o644);
+        builder.install(&src.join("LICENSE-MIT"), &doc, 0o644);
+        builder.install(&src.join("LICENSE-APACHE"), &doc, 0o644);
```
Comment (kennytm) [inline src/bootstrap/dist.rs]: Clippy doesn't have `LICENSE-MIT` + `LICENSE-APACHE`, it only has a `LICENSE` file.

**Review by kennytm — COMMENTED** [review]
LGTM except one item, but cc @Mark-Simulacrum for double-checking.

**Comment by o01eg** [comment]
Could you mention in `config.toml.example` that `clippy` is available in `build.tools` array?

File: src/bootstrap/dist.rs [hunk 3]
```diff
@@ -1181,6 +1183,83 @@ impl Step for Rls {
...
+        let version = builder.clippy_info.version(builder, &release_num);
+
+        let tmp = tmpdir(builder);
+        let image = tmp.join("clippy-image");
+        drop(fs::remove_dir_all(&image));
+        t!(fs::create_dir_all(&image));
+
+        // Prepare the image directory
+        // We expect clippy to build, because we've exited this step above if tool
+        // state for clippy isn't testing.
```
Comment (Mark-Simulacrum) [inline src/bootstrap/dist.rs]: er, I don't see where this happens above -- maybe the comment is stale?

File: src/bootstrap/dist.rs [hunk 4]
```diff
@@ -1181,6 +1183,83 @@ impl Step for Rls {
...
+        let release_num = builder.release_num("clippy");
+        let name = pkgname(builder, "clippy");
+        let version = builder.clippy_info.version(builder, &release_num);
+
+        let tmp = tmpdir(builder);
+        let image = tmp.join("clippy-image");
+        drop(fs::remove_dir_all(&image));
+        t!(fs::create_dir_all(&image));
+
+        // Prepare the image directory
```
Comment (Mark-Simulacrum) [inline src/bootstrap/dist.rs]: This also seems a bit too high.

File: src/bootstrap/dist.rs [hunk 5]
```diff
@@ -1608,6 +1712,9 @@ impl Step for Extended {
                 if rls_installer.is_some() {
                     cmd.arg("-dRlsDir=rls");
                 }
+                if clippy_installer.is_some() {
+                    cmd.arg("-dRlsDir=clippy");
```
Comment (Mark-Simulacrum) [inline src/bootstrap/dist.rs]: I'm not actually sure what is happening here, but should this perhaps be `-dClippyDir`?

File: src/bootstrap/dist.rs (hunk 2 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/dist.rs]: This still needs to be fixed.

**Comment by Mark-Simulacrum** [comment]
r=me pending @kennytm's nit

**Comment by oli-obk** [comment]
Oh yea, this PR is blocked on the next beta because we need to update the stage0 cargo

**Comment by oli-obk** [comment]
So... we got a new rustc bootstrap compiler. But cargo is still an old version. I should have probably backported the cargo submodule bump to beta -.-

Thus now we wait for the next bootstrap compiler bump

**Comment by Mark-Simulacrum** [comment]
We didn't actually get a new bootstrap compiler -- that PR was not intended for master. Will probably be ready Friday-ish.

**Comment by mati865** [comment]
@oli-obk last nightly is still `e3bf634e0 2018-06-28`

**Comment by mati865** [comment]
Well that doesn't look good.
Nightly release is blocked on `rls` update (after breaking https://github.com/rust-lang/rust/pull/51569), `rls` depends on `clippy` so it's blocked on bootstrap (relevant comment https://github.com/rust-lang/rust/pull/51677#discussion_r197325055).
In short `rls` is blocked on nightly release and nightly release is blocked on `rls` 🔁

**Comment by oli-obk** [comment]
why does rls depend on clippy?

**Comment by mati865** [comment]
@oli-obk it's built with `clippy_lints` enabled https://github.com/rust-lang/rust/pull/51677#issuecomment-399071418
I think disabling it (like I mentioned in rls PR) is best solution for now.

**Comment by oli-obk** [comment]
I'll just update clippy to work on master. Then we don't have any cycle issues.

**Comment by mati865** [comment]
Actually nightly builds have nothing to do with this PR, I must have confused things due to lateness.
Build failed with new beta as bootstrap:
```
[00:00:47] downloading https://static.rust-lang.org/dist/2018-06-30/rust-std-beta-x86_64-unknown-linux-gnu.tar.gz
[...]
[00:02:48] Caused by:
[00:02:48]   the cargo feature `edition` requires a nightly version of Cargo, but this is the `beta` channel
```

**Comment by oli-obk** [comment]
I created a branch on the clippy repo with a fix so it works on master

@bors r=Mark-Simulacrum p=1

**Comment by kennytm** [comment]
Legit.

```
duplicate artfacts found when compiling a tool, this typically means that something was recompiled because a transitive dependency has different features activated than in a previous build:
  num-traits 0.2.2 (registry+https://github.com/rust-lang/crates.io-index)
    `clippy-driver` enabled features [] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libnum_traits-8f3829bc1c5a543b.rlib"
    `cargo` enabled features ["default", "std"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libnum_traits-f9369633da61cbc8.rlib"
  serde_json 1.0.15 (registry+https://github.com/rust-lang/crates.io-index)
    `clippy-driver` enabled features ["default"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libserde_json-aa883cbbd3cf7a8a.rlib"
    `cargo` enabled features ["default"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libserde_json-6050d8d623a892c4.rlib"
  winapi 0.3.4 (registry+https://github.com/rust-lang/crates.io-index)
    `clippy-driver` enabled features ["consoleapi", "dbghelp", "errhandlingapi", "fileapi", "handleapi", "ioapiset", "minwindef", "namedpipeapi", "ntdef", "processenv", "processthreadsapi", "std", "winerror", "winnt", "winsock2", "ws2def", "ws2ipdef", "ws2tcpip"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libwinapi-cb891adfe3428d99.rlib"
    `cargo` enabled features ["consoleapi", "dbghelp", "errhandlingapi", "fileapi", "handleapi", "ioapiset", "jobapi", "jobapi2", "lmcons", "minschannel", "minwinbase", "minwindef", "namedpipeapi", "ntdef", "ntsecapi", "ntstatus", "processenv", "processthreadsapi", "profileapi", "psapi", "schannel", "securitybaseapi", "std", "synchapi", "sysinfoapi", "timezoneapi", "winbase", "wincon", "wincrypt", "winerror", "winnt", "winsock2", "ws2def", "ws2ipdef", "ws2tcpip"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libwinapi-a10c5bdd51d41051.rlib"
  backtrace 0.3.6 (registry+https://github.com/rust-lang/crates.io-index)
    `clippy-driver` enabled features ["backtrace-sys", "coresymbolication", "dbghelp", "default", "dladdr", "libbacktrace", "libunwind", "winapi"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libbacktrace-cbc02761c3050de3.rlib"
    `cargo` enabled features ["backtrace-sys", "coresymbolication", "dbghelp", "default", "dladdr", "libbacktrace", "libunwind", "winapi"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libbacktrace-4af4835397e19d58.rlib"
  error-chain 0.11.0 (registry+https://github.com/rust-lang/crates.io-index)
    `clippy-driver` enabled features ["backtrace", "default", "example_generated"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\liberror_chain-a916613177fdf7ce.rlib"
    `rls` enabled features ["backtrace", "default", "example_generated"] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\liberror_chain-68a9144955b21f58.rlib"
  cargo_metadata 0.5.4 (registry+https://github.com/rust-lang/crates.io-index)
    `clippy-driver` enabled features [] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libcargo_metadata-3fac61eab5a5a64c.rlib"
    `rls` enabled features [] at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libcargo_metadata-09486407d5b9061a.rlib"
```

**Comment by mati865** [comment]
Maybe clippy version used by rls gotta be the same as clippy component?

**Comment by oli-obk** [comment]
@bors r-

I only updated the LICENSE file mess for clippy docs, not clippy itself

**Comment by oli-obk** [comment]
@bors r+

passes locally now

**Comment by oli-obk** [comment]
No trumpets? Clippy appears still broken, it worked locally :/ Maybe another change jumped in

**Comment by o01eg** [comment]
Was it forgotten to add clippy to `Kind::Install` in `builder.rs`?

**Pull request merged**